File: src/chapters/numerical-calculus/content/code/gauss-quad.r
# 2- or 3-point Gauss-Legendre quadrature on [a, b].
gauss_quad <- function(f, a, b, n = 2) {
  if (n == 3) {
    t <- c(-sqrt(3 / 5), 0, sqrt(3 / 5)); w <- c(5 / 9, 8 / 9, 5 / 9)
  } else {
    t <- c(-1 / sqrt(3), 1 / sqrt(3)); w <- c(1, 1)
  }
  hm <- (b - a) / 2                 # map [-1,1] -> [a,b]
  mid <- (a + b) / 2
  hm * sum(w * f(mid + hm * t))
}

cat("int_0^1 e^x dx ~", gauss_quad(exp, 0, 1, 2), "(2-pt)\n")
cat("int_0^1 e^x dx ~", gauss_quad(exp, 0, 1, 3), "(3-pt)\n")
# -> 1.7178964 (2-pt), 1.7182810 (3-pt); exact e-1 = 1.7182818

File: src/chapters/numerical-calculus/content/code/gauss-quad.rs
// 2- or 3-point Gauss-Legendre quadrature on [a, b].
fn gauss_quad<F: Fn(f64) -> f64>(f: F, a: f64, b: f64, n: usize) -> f64 {
    let (t, w): (Vec<f64>, Vec<f64>) = if n == 3 {
        (vec![-(3.0_f64 / 5.0).sqrt(), 0.0, (3.0_f64 / 5.0).sqrt()], vec![5.0 / 9.0, 8.0 / 9.0, 5.0 / 9.0])
    } else {
        (vec![-1.0 / 3.0_f64.sqrt(), 1.0 / 3.0_f64.sqrt()], vec![1.0, 1.0])
    };
    let hm = (b - a) / 2.0;               // map [-1,1] -> [a,b]
    let mid = (a + b) / 2.0;
    hm * t.iter().zip(&w).map(|(&ti, &wi)| wi * f(mid + hm * ti)).sum::<f64>()
}
fn main() {
    println!("int_0^1 e^x dx = {} (2-pt)", gauss_quad(|x: f64| x.exp(), 0.0, 1.0, 2));
    println!("int_0^1 e^x dx = {} (3-pt)", gauss_quad(|x: f64| x.exp(), 0.0, 1.0, 3));
}
// -> 1.7178964 (2-pt), 1.7182810 (3-pt); exact e-1 = 1.7182818
